### ISSUE TITLE
JWT Signature Verification Fails with BadPaddingException Message is larger than modulus

### ISSUE THREAD

**Issue by andersonkyle** [spring-projects/spring-security-oauth#1052]
## Problem
Valid signed JWT tokens fail verification.  `BadPaddingException: Message is larger than modulus`

## Expectation
Valid signed JWT tokens should pass verification.

## Description
I'm using Okta as my Identity as a Service provider.  An upstream web application receives a JWT access token that is signed by Okta's public key in the JWK format.

**JWK Public Key**
```javascript
{
 "keys": [
  {
   "alg": "RS256",
   "e": "...",
   "n": "...",
   "kid": "...",
   "kty": "RSA",
   "use": "sig"
  }
 ]
}
```
I then pass that JWT access token to downstream API services.  This is where the failure occurs.  These API services fail to verify the signature of the JWT and throw the Exception mentioned above.

This shouldn't be the case.  I've tested a different JWT library and verified the signature successfully.

Tested with [nimbus-jose-jwt](https://developer.unboundid.com/6.0.0.0/broker/guides/broker-client-developer-guide/authentication/jwt/) with the following code:

```java
SignedJWT idToken = SignedJWT.parse("eyJhbGciOiJSUzI1NiJ9.eyJhdXRo...");
String kid = idToken.getHeader().getKeyID();

JWKSet jwks = JWKSet.load(new URL("https://example.com/jwks"));
RSAKey jwk = (RSAKey) jwks.getKeyByKeyId(kid);

JWSVerifier verifier = new RSASSAVerifier(jwk);
if (idToken.verify(verifier)) {
  System.out.println("valid signature");
} else {
  System.out.println("invalid signature");
}
```

So that begs the question of why does signature verification fail when using `spring-security-jwt`?

My API services are all Spring Boot apps and I've verified that it fetches the JWK public key successfully from this property `security.oauth2.resource.jwk.key-set-uri=...`

The [RsaVerifier](https://github.com/spring-projects/spring-security-oauth/blob/master/spring-security-jwt/src/main/java/org/springframework/security/jwt/crypto/sign/RsaVerifier.java#L48-L61) class from `spring-security-jwt` uses `java.security.Signature` which in turn leverages `sun.security.rsa.RSACore` which is ultimately where the Exception is thrown from.

**RSACore**
```java
private static BigInteger parseMsg(byte[] var0, BigInteger var1) throws BadPaddingException {
        BigInteger var2 = new BigInteger(1, var0);
        if(var2.compareTo(var1) >= 0) {
            throw new BadPaddingException("Message is larger than modulus");
        } else {
            return var2;
        }
    }
```
So I've identified where the Exception is thrown but not being a crypto expert, I'm hoping y'all can provide some guidance on this matter.  This is a blocker for me.

/cc @dsyer @tekul

**Comment by dsyer** [comment]
It's actually not all that clear what you are doing yet. The most likely explanation is an algorithm mismatch. It would be straightforward to debug if you had a sample token and public key.

**Comment by andersonkyle** [comment]
@dsyer thanks for the reply.  So I've been debugging this all morning and this is where I'm at.

The discrepancy I'm seeing between `spring-security-jwt` and `nimbus-jose-jwt` (randomly picked alternative library for sanity checking) is how they persist the `modulus` from the Public Key.

The modulus is persisted in the `RSAPublicKeyImpl` as a `BigInteger` field named `n`.

The String representation of the modulus, pulled from the public JWK JSON endpoint, is Base 64 decoded and saved as a `BigInteger`.

The Base 64 decoded values are the same for both libraries, but its the choice of constructors that appears to be different.

`spring-security-jwt` uses the single argument constructor for `BigInteger`

```java
public BigInteger(byte[] val) {
        if (val.length == 0)
            throw new NumberFormatException("Zero length BigInteger");

        if (val[0] < 0) {
            mag = makePositive(val);
            signum = -1;
        } else {
            mag = stripLeadingZeroBytes(val);
            signum = (mag.length == 0 ? 0 : 1);
        }
        if (mag.length >= MAX_MAG_LENGTH) {
            checkRange();
        }
    }
```

while `nimbus-jose-jwt` uses the two argument constructor
```java
public BigInteger(int signum, byte[] magnitude) {
        this.mag = stripLeadingZeroBytes(magnitude);

        if (signum < -1 || signum > 1)
            throw(new NumberFormatException("Invalid signum value"));

        if (this.mag.length == 0) {
            this.signum = 0;
        } else {
            if (signum == 0)
                throw(new NumberFormatException("signum-magnitude mismatch"));
            this.signum = signum;
        }
        if (mag.length >= MAX_MAG_LENGTH) {
            checkRange();
        }
    }
```
You can find the details on those two arguments in the [BigInteger](https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html) Javadoc.

This extra parameter has an effect on the value of the modulus and thus when `compareTo` is called from `RSACore.parseMsg` a different result occurs.  In the case of `spring-security-jwt` the comparison triggers a `BadPaddingException: Message is larger than modulus` Exception.

I've sent you an example JWT and Public Key by email.

Let me know if you concur with my findings.  Thanks!!

**Comment by andersonkyle** [comment]
Forgot to mention,

`nimbus-jose-jwt` uses the two argument constructor for `BigInteger` and provides a hard-coded value of 1 for the first parameter (signum)

**Comment by andersonkyle** [comment]
Looks like this exact issue was raised in #1010 and was fixed in [0651688](https://github.com/spring-projects/spring-security-oauth/commit/0651688c5ef21c85a3a58576380b04870c47b12c)

The fix was exactly what I thought, supplying the signum parameter when creating the BigInteger for the modulus.

Wish I had seen this sooner :disappointed:

Hope this release gets cut soon.  Thanks Dave.
